[clang] Fix -Wdouble-promotion in C++ list-initialization

mjacobse opened the pull request:
Resolves https://github.com/llvm/llvm-project/issues/33409.

The information `IsListInit` is already passed to function `CheckImplicitConversion` for another use-case which makes adding a condition for the double-promotion case simple.

Also adds tests, both for the changed list-initialization case as well as for normal explicit casts which already would have passed before this PR. These negative tests are added directly next to the positive tests in `warn-double-promotion.c` or for the C++-specific cases in a new .cpp version of that file.

mjacobse committed:
[clang] Fix -Wdouble-promotion in C++ list-init

A C++ list-initialization explicitly asks for the promotion to happen,
much like an explicit static_cast, so it should not be warned about.
Fixes #33409

mjacobse committed:
[clang] Test -Wdouble-promotion with explicit cast

Add tests to ensure that -Wdouble-promotion does not warn when promotion
is asked for explicitly by an explicit cast or C++ list initialization.
For the latter this creates a .cpp version of warn-double-promotion.c.
Test case for #33409

Sirraide reviewed (commented):
The implementation lgtm, but I think we can do some deduplication in the tests by adding a second `RUN` line to the `.c` file that specifies `-x c++` and then only adding tests for truly C++-specific constructs in the `.cpp` file.

Also, while you’re at it, could you add some tests for functional-style casts (e.g. `double(f)`) as well as some tests involving templates (e.g. `T{f}` where `T` is a template parameter instantiated as `double`).

mjacobse commented:
Thanks, your `struct B` example raises a very good point, because with this PR no warning would be thrown anymore for the promotion of `f` to `double` in the initialization of member `B::d`. Using parantheses
```c++
B(float f) : d(f) {}
```
would bring back the warning. Something similar happens with local variable initialization too:

```c++
void test(float f) {
  double d0 = f;
  double d1(f);
  double d2{f};
}
```

Before this PR, all three initializations give the warning. After this PR, `d0` and `d1` give the warning but `d2` doesn't.

I could see an argument for `d1` and `d2` not to issue a warning, but them behaving differently does seem wrong. For the member initializer list it seems more clear-cut that the warning is desirable, with the declaration as `double` and the initialization being split apart.

mjacobse committed:
Use long double alias to avoid compound literals

With (long double){f} we were getting the compound literal construction
from C99 instead of the intended C++ function-style cast with
list-initialization syntax.

mjacobse committed:
Avoid duplication of non-C++ specific tests

Run warn-double-promotion.c in both C and C++ mode and only add
those C++ tests with warn-double-promotion.cpp that are not
valid C.

mjacobse committed:
Add tests for function style casts

mjacobse committed:
Avoid unintended suppression of -Wdouble-promotion

This reverts commit 4f63f703b5bb56d2def94b99cd05212652605e42
which was the previous attempt to suppress -Wdouble-promotion when
explicitly asked for with C++ list initialization. Just checking if
we are somehow in a list initialization was way too loose and
suppressed many cases that should not have been.

The new fix works with the observation that the case of explicit
C++ list initialization in this case turns into a
CXXFunctionalCastExpr with InitListExpr as direct child. While the
CXXFunctionalCastExpr is ignored when checking for implicit
conversions, the InitListExpr is not, which causes the
-Wdouble-promotion warning. To fix this, treat the InitListExpr as
part of the CXXFunctionalCastExpr and skip it too.

mjacobse committed:
Add C++-specific init tests for -Wdouble-promotion

mjacobse commented:
Yeah that makes sense to me. So I reverted this attempt of just checking if we are in some kind of ListInit, it does not work.

It looks like expressions like `double{f}` turn into a `CXXFunctionalCastExpr` just like `double(f)`, with the only difference that for `double{f}` the `CXXFunctionalCastExpr` gets an additional direct child `InitListExpr` with one initializer. So the implicit conversion analysis does skip over `CXXFunctionalCastExpr` as an explicit cast, but not over the `InitListExpr`, causing the warning. But it seems like at least for this purpose, `InitListExpr` basically belongs to the explicit cast still, so I added some code to skip it too in this case. And this seems to work as intended.

The implementation is not the prettiest though and it seems to rely heavily on this implicit assumption that `InitListExpr` as a direct child of `CXXFunctionalCastExpr` means that we have this special case of list-initialization. I'd be very happy about suggestions for how to improve this.

I also applied the test deduplication that you suggested and added some tests for classic function-style casts and the initialization examples that showed that the previous fix was broken. I can add further suggested tests involving classes and templates later if the new fix turns out to be the right idea.

I also introduced an alias for `long double` in the tests to write `LongDouble{f}` instead of `(long double){f}`. The latter did not do what I expected, instead it results in a C99 compound-literal construction (which does trigger `-Wc99-extensions`). These actually produce the `-Wdouble-promotion` warning too while they probably shouldn't, but I suppose that's something for a different issue.

mjacobse committed:
Fix formatting

Sirraide reviewed (commented):
I’ve thought about this for a bit, and this approach seems fine to me

mjacobse committed:
Inline silencing of -Wdouble-promotion

To make it more clear what's going on

mjacobse committed:
Add -Wdouble-promotion tests with extra parens

mjacobse committed:
Add more specific C++ -Wdouble-promotion

Includes tests with classes and templates

mjacobse committed:
Fix grammar in comment

Co-authored-by: Sirraide <aeternalmail@gmail.com>

mjacobse committed:
Add -Wdouble-promotion test with extra braces

mjacobse commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -12797,6 +12797,22 @@ static void CheckCommaOperand(
     S.CheckImplicitConversion(E, T, CC);
 }
 
+static Expr *IgnoreExplicitCastForImplicitConversionCheck(ExplicitCastExpr *E) {
```
Added as suggested. Let me know if you'd like to see any more tests.

Sirraide reviewed (approved):
LGTM, thanks!
